### PR TITLE
Make ploting of reject_log more flexible.

### PR DESCRIPTION
**Pull request by mkoculak**
Hi,
I made a few changes that help creating plots of autoreject's reject_log in an automated pipeline.

1. I've added a parameter 'show' to allow suppression of plot display.
2. Added a check if the number of epochs exceeds the number of sensors. If that is the case, the axes are flipped and epochs are on x and sensors on y.
3. Made the function return the matplotlib figure, so it can be modified and saved.

### CODE REVIEW

**Comment by jasmainak** [comment]
looks reasonable to me, @agramfort ?

**Comment by dengemann** [comment]
Hey, thanks for the suggestions.
I am not so happy with 2. Perhaps an explicit option would be better than magically changing the behavior.
What do you think @jasmainak ?

**Comment by mkoculak** [comment]
You're probably right.
So maybe change it for a parameter 'horizontal' to do that flip explicitly?

File: autoreject/autoreject.py [hunk 1]
```diff
@@ -1218,26 +1218,51 @@ def __init__(self, bad_epochs, labels, ch_names):
         assert len(bad_epochs) == labels.shape[0]
         assert len(ch_names) == labels.shape[1]
 
-    def plot(self):
-        """Plot."""
+    def plot(self, horizontal=False, show=True):
```
Comment (jasmainak) [inline autoreject/autoreject.py]: I would call this parameter `orientation` which takes two params `'vertical'` and `'horizontal'`. To mimic matplotlib API as much as possible

**Comment by jasmainak** [comment]
@mkoculak you need to follow pep8, otherwise Travis will complain

File: autoreject/autoreject.py (hunk 1 above)
Comment (mkoculak) [reply autoreject/autoreject.py]: done!

File: autoreject/autoreject.py [hunk 2]
```diff
@@ -1218,26 +1218,55 @@ def __init__(self, bad_epochs, labels, ch_names):
...
+                     yticklabels=ch_names_)
+        elif orientation == 'vertical':
+            plt.imshow(self.labels, cmap='Reds',
+                       interpolation='nearest')
+            ax.set_xlabel('Channels')
+            ax.set_ylabel('Epochs')
+            plt.setp(ax, xticks=range(7, self.labels.shape[1], 10),
+                     xticklabels=ch_names_)
+        else:
+            msg = "Orientation can be only 'horizontal' or 'vertical'."
```
Comment (jasmainak) [inline autoreject/autoreject.py]:
```suggestion
            msg = "orientation can be only 'horizontal' or 'vertical'. Got %s" % orientation
```

Sorry, one last fix request!

File: autoreject/autoreject.py [hunk 3]
```diff
@@ -1218,26 +1218,55 @@ def __init__(self, bad_epochs, labels, ch_names):
...
+        """Plot.
+
+        Parameters
+        ----------
+        orientation : 'vertical' or 'horizontal'
+            If `'vertical'`, will plot sensors on x-axis and epochs on y-axis.
+            If `'horizontal'`, will plot epochs on x-axis and sensors
+            on y-axis.
+        show : bool
+            Display the figure immediately.
```
Comment (jasmainak) [inline autoreject/autoreject.py]:
```suggestion
            If True, display the figure immediately.
```

File: autoreject/autoreject.py [hunk 4]
```diff
@@ -1218,26 +1218,56 @@ def __init__(self, bad_epochs, labels, ch_names):
...
+        figure : Instance of matplotlib.figure.Figure
+        """
         import matplotlib.pyplot as plt
 
-        plt.figure(figsize=(12, 6))
-        plt.imshow(self.labels, cmap='Reds',
-                   interpolation='nearest')
-        # XXX to be fixed
-        ch_names_ = self.ch_names[7::10]
+        figure = plt.figure(figsize=(12, 6))
```
Comment (agramfort) [inline autoreject/autoreject.py]:
```suggestion
        figure, ax = plt.subplots(figsize=(12, 6))
```

then no need for plt.gca()

File: autoreject/autoreject.py [hunk 5]
```diff
@@ -1218,26 +1218,56 @@ def __init__(self, bad_epochs, labels, ch_names):
...
         ax.grid(False)
-        ax.set_xlabel('Channels')
-        ax.set_ylabel('Epochs')
-        plt.setp(ax, xticks=range(7, self.labels.shape[1], 10),
-                 xticklabels=ch_names_)
+        ch_names_ = self.ch_names[7::10]
+
+        if orientation == 'horizontal':
+            plt.imshow(self.labels.T, cmap='Reds',
+                       interpolation='nearest')
```
Comment (agramfort) [inline autoreject/autoreject.py]: ax.imshow

File: autoreject/autoreject.py [hunk 6]
```diff
@@ -1218,26 +1218,56 @@ def __init__(self, bad_epochs, labels, ch_names):
...
+            ax.set_xlabel('Epochs')
+            ax.set_ylabel('Channels')
+            plt.setp(ax, yticks=range(7, self.labels.shape[1], 10),
```
Comment (agramfort) [inline autoreject/autoreject.py]: ax.setp?

File: autoreject/autoreject.py [hunk 7]
```diff
@@ -1218,26 +1218,56 @@ def __init__(self, bad_epochs, labels, ch_names):
...
-        ax.set_ylabel('Epochs')
-        plt.setp(ax, xticks=range(7, self.labels.shape[1], 10),
-                 xticklabels=ch_names_)
+        ch_names_ = self.ch_names[7::10]
+
+        if orientation == 'horizontal':
+            plt.imshow(self.labels.T, cmap='Reds',
+                       interpolation='nearest')
+            ax.set_xlabel('Epochs')
+            ax.set_ylabel('Channels')
```
Comment (agramfort) [inline autoreject/autoreject.py]: xylabels code setting is duplicated

File: autoreject/autoreject.py (hunk 4 above)
Comment (mkoculak) [reply autoreject/autoreject.py]: right, done.

File: autoreject/autoreject.py (hunk 5 above)
Comment (mkoculak) [reply autoreject/autoreject.py]: done.

File: autoreject/autoreject.py (hunk 6 above)
Comment (mkoculak) [reply autoreject/autoreject.py]: doesn't work.

File: autoreject/autoreject.py (hunk 7 above)
Comment (mkoculak) [reply autoreject/autoreject.py]: the x and y labels are flipped, not duplicated.

Comment (agramfort) [reply autoreject/autoreject.py]: I mean you use this in the 2 if conditions blocks. See below. Can these 2 lines after the if.

Comment (jasmainak) [reply autoreject/autoreject.py]: @agramfort on L1246, it is:

```py
ax.set_xlabel('Epochs')
```

and L1253, it is:

```py
ax.set_xlabel('Channels')
```

They are different even though they look similar.

**Comment by agramfort** [comment]
indeed I had read too quickly. Can you update a test to avoid reducing code coverage?

**Comment by mkoculak** [comment]
Would love to, but honestly, I have never done that before. So I would be grateful for any hints.
Also, I'm not sure if I am looking at the correct file, but I don't see any test for the plotting function.

**Comment by jasmainak** [comment]
I would add a line after this: https://github.com/autoreject/autoreject/blob/ae2733e9e6408e43ded3c78277d584efde925b75/autoreject/tests/test_viz.py#L41

Not sure why `reject_log.plot` is not called, but you would essentially need to call it and check that it works for the two arguments but not for a random string.

**Comment by mkoculak** [comment]
Is this what you had in mind?

**Comment by jasmainak** [comment]
@mkoculak excellent! The test looks great. Can you record your contribution in the `whats_new.rst` page? It will be part of the release notes for the next release.

**Comment by mkoculak** [comment]
Done. Never used rst before, but I hope I inserted it correctly.

File: doc/whats_new.rst [hunk 8]
```diff
@@ -15,6 +15,8 @@ Current
 Changelog
 ~~~~~~~~~
 
+- Make plotting of reject log more flexible by allowing to change the orientation of plot, option to supress the immediate display of the plot and returing the plot figure for further modification or saving in :meth:`autoreject.RejectLog.plot`, by `Marcin Koculak`_ in `#152 <https://github.com/autoreject/autoreject/pull/152>`_
```
Comment (agramfort) [inline doc/whats_new.rst]:
```suggestion
- Make plotting of reject log more flexible by allowing to change the orientation of plot, option to supress the immediate display of the plot and returning the plot figure for further modification or saving in :meth:`autoreject.RejectLog.plot`, by `Marcin Koculak`_ in `#152 <https://github.com/autoreject/autoreject/pull/152>`_
```

**Comment by jasmainak** [comment]
@mkoculak fix the typo that @agramfort pointed out and I'll merge!

**Comment by agramfort** [comment]
@jasmainak merge if you're happy

**Comment by mkoculak** [comment]
Thanks for all the help! I learnt quite a bit about github from this experience. :)

**Comment by jasmainak** [comment]
Merged, thanks @mkoculak for your contribution and glad to know you enjoyed the review process!

**Pull request merged**